pkg/vsphere/simulator/simulator.go

```diff
@@ -0,0 +1,236 @@
+	"github.com/vmware/govmomi/vim25/xml"
+	"github.com/vmware/vic/pkg/vsphere/simulator/esx"
+	"github.com/vmware/vic/pkg/vsphere/simulator/vc"
+)
+
+// Method encapsulates a decoded SOAP client request
+type Method struct {
+	Name string
+	Body interface{}
+}
+
+// Handler implements a response handler for a Method request
+type Handler interface {
+	Name() string
+	Call(*Method) soap.HasFault
+}
+
+// Service decodes incoming requests and dispatches to a Handler
+type Service struct {
+	handlers map[string]Handler
+
+	readAll func(io.Reader) ([]byte, error)
+}
+
+// Server provides a simulator Service over HTTP
+type Server struct {
+	*httptest.Server
+	URL *url.URL
+}
+
+// New returns an initialized simulator Service instance
+func New() *Service {
+	s := &Service{
+		handlers: make(map[string]Handler),
+
+		readAll: ioutil.ReadAll,
+	}
+
+	return s
+}
+
+// Register method handlers with a Service
+func (s *Service) Register(handlers ...Handler) {
+	for _, handler := range handlers {
+		s.handlers[handler.Name()] = handler
+	}
+}
+
+var handlers = []Handler{
+	&Login{},
+	&CurrentTime{},
+}
+
+// AsESX registers service handlers with an ESX personality
+func (s *Service) AsESX() *Service {
+	h := []Handler{
+		&RetrieveServiceContent{esx.ServiceContent},
+	}
+	s.Register(append(handlers, h...)...)
+	return s
+}
+
+// AsVC registers service handlers with a vCenter personality
+func (s *Service) AsVC() *Service {
+	h := []Handler{
+		&RetrieveServiceContent{vc.ServiceContent},
+	}
+	s.Register(append(handlers, h...)...)
+	return s
+}
+
+type serverFaultBody struct {
+	Reason *soap.Fault `xml:"http://schemas.xmlsoap.org/soap/envelope/ Fault,omitempty"`
+}
+
+func (b *serverFaultBody) Fault() *soap.Fault { return b.Reason }
+
+func serverFault(msg string) soap.HasFault {
+	return &serverFaultBody{Reason: Fault(msg, &types.InvalidRequest{})}
+}
+
+// Fault wraps the given message and fault in a soap.Fault
+func Fault(msg string, fault types.BaseMethodFault) *soap.Fault {
+	f := &soap.Fault{
+		Code:   "ServerFaultCode",
+		String: msg,
+	}
+
+	f.Detail.Fault = fault
+
+	return f
+}
+
+// ServeHTTP implements the http.Handler interface
+func (s *Service) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	if r.Method != "POST" {
+		w.WriteHeader(http.StatusMethodNotAllowed)
+		return
+	}
+
+	body, err := s.readAll(r.Body)
+	_ = r.Body.Close()
+	if err != nil {
+		log.Printf("error reading body: %s", err)
+		w.WriteHeader(http.StatusBadRequest)
+		return
+	}
+
+	var res soap.HasFault
+
+	method, err := UnmarshalBody(body)
+	if err != nil {
+		res = serverFault(err.Error())
+	} else {
+		if handler, ok := s.handlers[method.Name]; ok {
+			res = handler.Call(method)
+		} else {
+			res = serverFault(fmt.Sprintf("no handler for %s", method.Name))
+		}
+	}
+
+	if res.Fault() == nil {
+		w.WriteHeader(http.StatusOK)
+	} else {
+		w.WriteHeader(http.StatusInternalServerError)
+	}
+
+	e := xml.NewEncoder(w)
+	err = e.Encode(&soap.Envelope{Body: res})
+
```

Comment:
Fixed, along with another stray blank line in simulator_test.go.  I'll wait to merge until CI passes.